src/bipart.cpp

```diff
@@ -0,0 +1,535 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+// libsemigroups headers
+#include <libsemigroups/bipart.hpp>
+
+// pybind11....
+#include <pybind11/operators.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+// libsemigroups_pybind11....
+#include "main.hpp"  // for init_blocks, init_bipart
+
+namespace py = pybind11;
+
+namespace libsemigroups {
+
+  void init_blocks(py::module& m) {
+    py::class_<Blocks> thing(m,
+                             "Blocks",
+                             R"pbdoc(
+A :any:`Blocks` object represents a signed partition of the set :math:`\{0,
+\ldots, n - 1\}`.
+
+It is possible to associate to every :any:`Bipartition` a pair of blocks,
+:any:`Bipartition.left_blocks()` and :any:`Bipartition.right_blocks()` ,
+which determine the Green's :math:`\mathscr{L}`- and
+:math:`\mathscr{R}`-classes of the :any:`Bipartition` in the monoid of all
+bipartitions. This is the purpose of this class.
+)pbdoc");
+    thing.def("__repr__",
+              py::overload_cast<Blocks const&>(&to_human_readable_repr));
+    thing.def(py::self != py::self);
+    thing.def(py::self < py::self);
+    thing.def(py::self == py::self);
+
+    thing.def("__copy__", [](Blocks const& self) { return Blocks(self); });
+    thing.def(
+        "copy",
+        [](Blocks const& self) { return Blocks(self); },
+        R"pbdoc(
+Copy a Blocks object.
+
+:returns: A copy.
+:rtype: Blocks
+)pbdoc");
+
+    // We omit the other constructors because in python we can't do anything
+    // with them.
+
+    thing.def(py::init<std::vector<std::vector<int32_t>> const&>(),
+              py::arg("blocks"),
+              R"pbdoc(
+Constructs a Blocks object from a list of lists of integers.
+
+This function constructs a :any:`Blocks` object from a list of lists of
+integers, so that the blocks consisting of negative values are
+transverse and those consisting of positive values are not.
+
+:param blocks: the blocks
+:type blocks: list
+
+:raises LibsemigroupsError:
+  if the set consisting of the absolute values of the entries in ``blocks`` is
+  not :math:`\{1, \ldots, n\}` where :math:`n` is the maximum such value.
+
+:raises LibsemigroupsError:  if ``0`` is an item in any block.
+
+:raises LibsemigroupsError:  if any block is empty.
+
+:raises LibsemigroupsError:  if any block contains both negative and positive values.
+
+:raises LibsemigroupsError:  if the constructed :any:`Blocks` object is not valid.
+
+:complexity:
+  linear in the sum of the sizes of the vectors in blocks.
+)pbdoc");
+    thing.def(
+        "iterator",
+        [](Blocks const& self) {
+          return py::make_iterator(self.cbegin(), self.cend());
+        },
+        R"pbdoc(
+Return a const iterator pointing to the index of the first block.
+
+:complexity:
+   Constant.
+
+:returns:
+   A value of type :any:`Iterator`.
+
+:rtype:
+   Iterator
```

Comment:
```suggestion
   iterator
```

Reply:
Are you sure about this? If we use `Iterator` it links to the relevant part of the python docs.

Reply:
That link isn't being created for me, but nor is it being created when I use `iterator`. Since `iterator` isn't a type in python, I suppose it doesn't really matter. The actual return type seems to be `pybind11_builtins.iterator`.

Reply:
Ok, I'll go with `iterator` thanks!